### ISSUE TITLE
PermissionError(13, 'Access is denied.', None, 5) (ESPTOOL-334)

### ISSUE THREAD

**Issue by octolance** [espressif/esptool#682]
# Operating system
Microsoft Windows® 10 Version 21H1 Build 19043.1288
# Python version
Python 3.9.7

# What Chip
ESP8266

# What development board or other hardware is the chip attached to
NodeMCU 1.0 ESP-12E

# Is anything else attached to the development board, except for the serial flasher connections?
No

# Are you running esptool.py from an IDE such as Arduino or Eclipse?
No IDE, Windows command prompt

# Full esptool.py command line that was run:
`py -m esptool --port COM5 read_mac`

# Full output from esptool.py
```
esptool.py v3.1
Serial port COM5
Traceback (most recent call last):
  File "C:\Program Files\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Program Files\Python39\lib\site-packages\esptool.py", line 4582, in <module>
    _main()
  File "C:\Program Files\Python39\lib\site-packages\esptool.py", line 4575, in _main
    main()
  File "C:\Program Files\Python39\lib\site-packages\esptool.py", line 4074, in main
    esp = esp or get_default_connected_device(ser_list, port=args.port, connect_attempts=args.connect_attempts,
  File "C:\Program Files\Python39\lib\site-packages\esptool.py", line 116, in get_default_connected_device
    _esp = ESPLoader.detect_chip(each_port, initial_baud, before, trace,
  File "C:\Program Files\Python39\lib\site-packages\esptool.py", line 354, in detect_chip
    detect_port = ESPLoader(port, baud, trace_enabled=trace_enabled)
  File "C:\Program Files\Python39\lib\site-packages\esptool.py", line 313, in __init__
    self._port = serial.serial_for_url(port)
  File "C:\Users\Alex\AppData\Roaming\Python\Python39\site-packages\serial\__init__.py", line 90, in serial_for_url
    instance.open()
  File "C:\Users\Alex\AppData\Roaming\Python\Python39\site-packages\serial\serialwin32.py", line 64, in open
    raise SerialException("could not open port {!r}: {!r}".format(self.portstr, ctypes.WinError()))
serial.serialutil.SerialException: could not open port 'COM5': PermissionError(13, 'Access is denied.', None, 5)
```

# Do you have any other information from investigating this?
This problem is not present in `esptool.py v3.0` (a compiled binary was used)
The problem is not present when  not specifying the COM port i.e. `py -m esptool read_mac`. *esptool.py* successfully identifies the NodeMCU on COM5 and reads the MAC address successfully.

**Comment by ErikRust** [comment]
In my case I reinstalled drivers from https://sparks.gogo.co.nz/assets/_site_/downloads/CH34x_Install_Windows_v3_4.zip with Admin Rights, seeing as permission access is in question. This is on a windows system, so right click run as admin seem to have solved issue.

**Comment by Sawadee2u** [comment]
This worked for me (ErikRust) updated CH34x driver. 
USB-Serial CH340 Version 3.4.2014.8 (8/8/2014) Did not work and caused this error.
USB-Serial CH340 Version 3.5.2019.1 (1/30/2019) Worked with no errors.
Device used : n/dchart's USB-C to UPDI Key with CH340N chip 3 pogo pin flashlight programmer

**Comment by eagl1** [comment]
ErikRust, thank you so much my ESP32-8048S043C was struggling to connect to the serial port. Following your solution solved the problem.

**Comment by alain110** [comment]
Thank you very much, it works well now for me too.

**Comment by t0mkee** [comment]
> In my case I reinstalled drivers from https://sparks.gogo.co.nz/assets/_site_/downloads/CH34x_Install_Windows_v3_4.zip with Admin Rights, seeing as permission access is in question. This is on a windows system, so right click run as admin seem to have solved issue.

i have been trying for 4 hours trying to get my chip to work, THANK YOU

**Comment by eagl1** [comment]
Other settings I recommend:
1. Try installing newer or older CH340 drivers.
2. Enable --> Erase all flash before sketch upload.
3. Press the `boot` button before plugging in the board or with pressing the `rst` button.

**Comment by Cat-100** [comment]
I had the same issue... 

In device manager I disabled the drive, restarted the pc, and then enabled the drive again. It still didn't work. I fixed it by following the esp troubleshooting documentation [esp32 troubleshooting](https://docs.espressif.com/projects/esptool/en/latest/esp32/troubleshooting.html)

It looks like you need to remove any connected devices (wired) and then flash again and or attempt to access the COMx port. 

In summary:

- Remove connected Wiring to the esp32
- Attempting flashing again

That was the only thing that worked for me and I have been struggling with this for about 4 days, I hope this helps.

**Comment by riwalker** [comment]
See my fix [here](https://stackoverflow.com/questions/34550437/serialexception-could-not-open-port-access-is-denied/79403427#79403427)
Killing the stray 'PYTHON' task works great.

**Comment by Cyrillinux** [comment]
Hello,
I have two solutions for you.

1/ Try new package 3.1.3:
https://bacciel.com/tools/package_champollion_esp8266_index.json

2/ In a command prompt (cmd), use the command : serialinit.dll COM3 115200

(COM3 is an example; replace it with the correct port) before to use Arduino IDE, or just before upload the program to the ESP8266 :

https://sourceforge.net/projects/serialinit/

Download Serial Init Fix Windows 11 Error Init for free. Serial Init Fix Windows 11 Error Init. Serial Init Fix Windows 11 Error Init Prepare and configure serial port before use. Using : serialinit.dll COMx [baudrate] Baudrate : 9600, 19200,…

I’am using seconde solution for now, and work it fine.

See you soon. Good luck
Cyril

**Comment by Cyrillinux** [comment]
Hello, a another better solution,

try : CH34xFIX fix this error (Permission 13) : https://github.com/Cyrillinux/esptool/releases/download/packages/CH34xFIX.zip

Good luck

Cyril

**Comment by yutizf** [comment]
> In my case I reinstalled drivers from https://sparks.gogo.co.nz/assets/_site_/downloads/CH34x_Install_Windows_v3_4.zip with Admin Rights, seeing as permission access is in question. This is on a windows system, so right click run as admin seem to have solved issue.

Thnk, its work for me.

**Comment by heX16** [comment]
> Hello, a another better solution,
> 
> try : CH34xFIX fix this error (Permission 13) : https://github.com/Cyrillinux/esptool/releases/download/packages/CH34xFIX.zip
> 
> Good luck
> 
> Cyril

Thank you! It works for me! Working driver: version 3.3.2011.11, date 2011-11-04.

The issue i had: `Cannot configure port, something went wrong. Original message: PermissionError(13, 'A device attached to the system is not functioning.', None, 31)`.

**Comment by owl5053** [comment]
> See my fix [here](https://stackoverflow.com/questions/34550437/serialexception-could-not-open-port-access-is-denied/79403427#79403427) Killing the stray 'PYTHON' task works great.

God bless you man :) win 25h2 - same issue. Fixed with ur IN

**Comment by LowPowerLab** [comment]
Access Denied ... what a dumb error. 
In my case it was somehow the USB-C cable that was the culprit although it makes no sense. Changing to another cable solved this.

I had the Windows default CH34x driver 3.9.2024.9 which worked fine uploading from Arduino IDE before running into this in VS/PIO thinking somehow VS/PIO has a problem with that driver. So I replaced it with CH34x 3.4. Still didn't work until I swapped the cable as last resort and voila, it worked. FWIW I reverted back to the 3.9.2024.9 and it all works just fine, so it was not a driver issue, just a stupid cable.

At the very least I'd recommend folks install drivers/executables from at least half decent reputable source. [Such as Sparkfun CH34x driver install tutorial](https://learn.sparkfun.com/tutorials/how-to-install-ch340-drivers/all).
